### PR TITLE
Clean Ceg Instantiators

### PR DESCRIPTION
**Pull request by ajreynol**
Also add infrastructure for theory-specific preprocessing in CEGQI.

This is in preparation for a preprocess phase for bit-vectors which sets up solving for disjoint bit-vector extract terms.

### CODE REVIEW

File: src/theory/quantifiers/ceg_instantiator.h [hunk 1]
```diff
@@ -138,12 +143,125 @@ class SolvedForm {
...
+   */
+  void presolve(Node q);
+  /** Register the counterexample lemma
+   *
+   * lems contains the conjuncts of the counterexample lemma of the quantified
+   *   formula we are processing. The counterexample lemma is the formula
+   *   { ~phi[e/x] } in Figure 1 of Reynolds et al. FMSD 2017.
+   * ce_vars contains the variables e. Notice these are variables of
+   *   INST_CONSTANT kind, since we do not permit bound variables in
+   *   assertions.
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_instantiator.h]: Is the indent here on purpose?

File: src/theory/quantifiers/ceg_instantiator.h [hunk 2]
```diff
@@ -168,25 +288,88 @@ class CegInstantiator {
...
+   * This recursively calls registerTheoryId for typeOf(tn') for
+   * all parameters and datatype subfields of type tn.
+   * visited stores the types we have already visited.
+   */
+  void registerTheoryIds(TypeNode tn, std::map<TypeNode, bool>& visited);
+  /** register theory id tid
+   *
+   * This is called when the quantified formula we are processing
+   * with this class involves theory tid. In this case, we will
+   * based instantiations based on the assertion list of this theory.
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_instantiator.h]: wording?

**Comment by ajreynol** [comment]
Yes, ceg_t_instantiator needs further cleanup, I've opened #1367.

File: src/theory/quantifiers/ceg_instantiator.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/ceg_instantiator.h]: it was although perhaps this looks strange in this case, I'll fix.

**Comment by ajreynol** [comment]
I've addressed the review in my latest commit.

**Pull request merged**